### PR TITLE
[AIRFLOW-6873] fix mypyd errors and add new type annotations

### PR DESCRIPTION
**Pull request by houqp**
fix mypyd type errors and add new type annotations

---
Issue link: [AIRFLOW-6873](https://issues.apache.org/jira/browse/AIRFLOW-6873)

Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Commit message/PR title starts with `[AIRFLOW-NNNN]`. AIRFLOW-NNNN = JIRA ID<sup>*</sup>
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

<sup>*</sup> For document-only changes commit message can start with `[AIRFLOW-XXXX]`.

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

**Comment by potiuk** [comment]
Looks like a real issue introduced somewhere :) - the three Travis builds are failing in similar way

File: airflow/jobs/scheduler_job.py [hunk 1]
```diff
@@ -631,7 +631,7 @@ def create_dag_run(self, dag, session=None):
                 return next_run
 
     @provide_session
-    def _process_task_instances(self, dag, task_instances_list, session=None):
+    def _process_task_instances(self, dag, task_instances_list: List[TaskInstanceKeyType], session=None):
```
Comment (turbaszek) [inline airflow/jobs/scheduler_job.py]:
```suggestion
    def _process_task_instances(self, dag: DAG, task_instances_list: List[TaskInstanceKeyType], session=None):
```

File: airflow/jobs/scheduler_job.py [hunk 2]
```diff
@@ -672,7 +672,7 @@ def _process_task_instances(self, dag, task_instances_list, session=None):
                     self.log.debug('Queuing task: %s', ti)
                     task_instances_list.append(ti.key)
 
-    def _process_dags(self, dagbag, dags, tis_out):
+    def _process_dags(self, dagbag, dags, tis_out: List[TaskInstanceKeyType]):
```
Comment (turbaszek) [inline airflow/jobs/scheduler_job.py]:
```suggestion
    def _process_dags(self, dagbag: DagBag, dags: List[DAG], tis_out: List[TaskInstanceKeyType]):
```

File: airflow/models/dag.py [hunk 3]
```diff
@@ -902,11 +902,11 @@ def topological_sort(self, include_subdag_tasks: bool = False):
     @provide_session
     def set_dag_runs_state(
             self,
-            state=State.RUNNING,
+            state: str = State.RUNNING,
             session=None,
             start_date=None,
             end_date=None,
```
Comment (turbaszek) [inline airflow/models/dag.py]: Can you add rest of the annotations?

File: airflow/models/dagrun.py [hunk 4]
```diff
@@ -349,9 +345,14 @@ def update_state(self, session=None):
 
         return ready_tis
 
-    def _get_ready_tis(self, scheduleable_tasks, finished_tasks, session):
+    def _get_ready_tis(
+        self,
+        scheduleable_tasks: List[TI],
+        finished_tasks: List[TI],
+        session,
```
Comment (turbaszek) [inline airflow/models/dagrun.py]: from sqlalchemy.orm.session import Session
```suggestion
        session: Session,
```

**Review by turbaszek — COMMENTED** [review]
Would you mind adding missing type annotations?

**Comment by houqp** [comment]
@potiuk @nuclearpinguin all feedback addressed, ready for another round of review.

File: airflow/models/dagbag.py [hunk 5]
```diff
@@ -116,7 +116,7 @@ def size(self):
     def dag_ids(self):
```
Comment (turbaszek) [inline airflow/models/dagbag.py]:
```suggestion
    def dag_ids(self) -> List[str]:
```

File: airflow/models/dagrun.py [hunk 6]
```diff
@@ -377,8 +378,13 @@ def _get_ready_tis(self, scheduleable_tasks, finished_tasks, session):
 
         return ready_tis, changed_tis
 
-    def _are_premature_tis(self, unfinished_tasks, finished_tasks, session):
-        # there might be runnable tasks that are up for retry and from some reason(retry delay, etc) are
+    def _are_premature_tis(
+        self,
+        unfinished_tasks: List[TI],
+        finished_tasks: List[TI],
+        session,
```
Comment (turbaszek) [inline airflow/models/dagrun.py]:
```suggestion
        session: Session,
```

File: airflow/utils/helpers.py [hunk 7]
```diff
@@ -56,7 +57,7 @@ def validate_key(k, max_length=250):
         return True
 
 
-def alchemy_to_dict(obj):
+def alchemy_to_dict(obj) -> Optional[Dict]:
```
Comment (turbaszek) [inline airflow/utils/helpers.py]:
```suggestion
def alchemy_to_dict(obj: Any) -> Optional[Dict]:
```

**Comment by houqp** [comment]
@nuclearpinguin updated.

**Review by turbaszek — APPROVED** [review]

**Comment by houqp** [comment]
@potiuk @ashb @nuclearpinguin all tests have passed, ready to merge?

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Thanks @houqp !

**Pull request merged**